libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt

```diff
@@ -69,39 +69,87 @@ object ClasspathChangesComputer {
         for (key in current.classSnapshots.keys) {
             val currentSnapshot = current.classSnapshots[key]!!
             val previousSnapshot = previous.classSnapshots[key] ?: return Failure.AddedRemovedClasses
-            if (currentSnapshot !is KotlinClassSnapshot || previousSnapshot !is KotlinClassSnapshot) {
-                return Failure.NotYetImplemented
+            val result = collectClassChanges(currentSnapshot, previousSnapshot, changesCollector)
+            if (result !is Success) {
+                return result
             }
-            // TODO: Store results in changesCollector
-            collectKotlinClassChanges(currentSnapshot, previousSnapshot)
         }
         return Success
     }
 
-    @TestOnly
-    internal fun collectKotlinClassChanges(current: KotlinClassSnapshot, previous: KotlinClassSnapshot): DirtyData {
+    private fun collectClassChanges(
+        current: ClassSnapshot,
+        previous: ClassSnapshot,
+        @Suppress("UNUSED_PARAMETER") changesCollector: ChangesCollector
+    ): ChangesCollectorResult {
+        if (current is JavaClassSnapshot && previous is JavaClassSnapshot &&
+            (current !is PlainJavaClassSnapshot || previous !is PlainJavaClassSnapshot)
+        ) {
+            return Failure.NotYetImplemented
+        }
+        // TODO: Store results in changesCollector and return SUCCESS here
+        computeClassChanges(current, previous)
+        return Failure.NotYetImplemented
+    }
+
+    @VisibleForTesting
+    internal fun computeClassChanges(current: ClassSnapshot, previous: ClassSnapshot): DirtyData {
         // TODO Create IncrementalJvmCache early once and reuse it here
         val workingDir =
             FileUtil.createTempDirectory(this::class.java.simpleName, "_WorkingDir_${UUID.randomUUID()}", /* deleteOnExit */ true)
         val incrementalJvmCache = IncrementalJvmCache(workingDir, /* targetOutputDir */ null, FileToCanonicalPathConverter)
+        val changesCollector = ChangesCollector()
 
+        when {
+            current is KotlinClassSnapshot && previous is KotlinClassSnapshot ->
+                collectKotlinClassChanges(current, previous, incrementalJvmCache, changesCollector)
+            current is JavaClassSnapshot && previous is JavaClassSnapshot ->
+                collectJavaClassChanges(current, previous, incrementalJvmCache, changesCollector)
+            else -> error("Incompatible types: ${current.javaClass.name} vs. ${previous.javaClass.name}")
```

Comment:
Yes, we should not fail here (I had planned to handle this case in the next PR). Added a TODO here.

Reply:
Actually it is a bit tricky. It will require taking both a Kotlin snapshot and a Java snapshot of a Kotlin class, so that we can fall back to comparing Java snapshots if one of the two (current and previous) snapshots is Kotlin and the other is Java. We can discuss this more in the next PR.

Reply:
We can consider all symbols defined in previous and current snapshot dirty, it can be overly broad, but it is safe and simple.

Reply:
Yup, that works. I'll do that.